### PR TITLE
51928 Add isFullRevision property to CDTDocumentRevision and subclasses

### PR DESCRIPTION
**Pull request by mikerhodes**
_Why_

Some operations, for example querying with a field filter, produce document revision objects which are not complete representations of the revision. In order to avoid data loss in applications, these revisions should not be allowed to be saved by the datastore.

_How_

https://github.com/cloudant/CDTDatastore/commit/0c722a5eeaa3e1d171284cc5029929e922cfcd87

Adds `isFullRevision` property to `CDTDocumentRevision`. This allows partial revision objects, such as `CDTQProjectedDocumentRevision`, to be recognised when a developer attempts to write one to the database. `-updateDocumentFromRevision:error:` uses `isFullRevision` to reject updates where the property is `NO`.

- `CDTDocumentRevision` sets the property to `YES` so it can be saved.
- `CDTProjectedDocumentRevision` sets the property to `NO` so it cannot be saved.
- `-updateDocumentFromRevision:error:` first checks the property before saving.

https://github.com/cloudant/CDTDatastore/commit/f6c8250797d9dded092f5dabbf486e8d9b9f6044

Re-enables tests which needed this functionality.

### CODE REVIEW

File: Classes/common/CDTDocumentRevision.h [hunk 1]
```diff
@@ -19,12 +19,14 @@
...
-@class CDTDocumentRevision;
 
 /**
  * Represents a single revision of a document in a datastore.
  */
-@interface CDTDocumentRevision : NSObject <CDTChangedObserver>
+@interface CDTDocumentRevision : NSObject <CDTChangedObserver> {
+   @protected
+    bool _isFullRevision;
+}
```
Comment (mikerhodes) [inline Classes/common/CDTDocumentRevision.h]: The `@protected` is required so sub-classes can set the property at `-init` time.

**Comment by tomblench** [comment]
Two thoughts on this:
- Could you swap the logic round and call the property `isProjectedRevision`? _or_
- Could you make better use of the class hierarchy, perhaps override `updateDocumentFromRevision` in `CDTQProjectedDocumentRevision` to throw the error there instead?

I guess it partly depends on whether you can envisage any other situations where you would encounter a "non-full/partial" revision.

**Comment by mikerhodes** [comment]
> - Could you swap the logic round and call the property isProjectedRevision?

From `updateDocumentFromRevision`'s perspective, it doesn't care why the revision isn't a complete revision. It's acting on the fact that the revision isn't a full one, not that it's projected. As you mentioned below, this property can be reused by other classes too, if need be.

> - Could you make better use of the class hierarchy, perhaps override updateDocumentFromRevision in CDTQProjectedDocumentRevision to throw the error there instead?

I think you're confusing `CDTDatastore` with `CDTDocumentRevision` here?

**Comment by tomblench** [comment]
Right, i got a bit confused about the classes involved. How about an `isFullRevision` method which returns `YES` for the non-projected version and `NO` for the projected one?

**Comment by mikerhodes** [comment]
@tomblench 
> Right, i got a bit confused about the classes involved. How about an isFullRevision method which returns YES for the non-projected version and NO for the projected one?

Isn't this basically what we have?

**Comment by tomblench** [comment]
+1

**Comment by rhyshort** [comment]
+1

**Pull request merged**